src/cpp/core/system/PosixSystem.cpp

```diff
@@ -546,36 +632,107 @@ Error closeNonStdFileDescriptors()
+   if (!error)
+   {
+      for (unsigned int fd : fds)
+      {
+
+         error = posixCall<std::size_t>(boost::bind(::write,
+                                                    pipeFd,
+                                                    &fd,
+                                                    4),
+                                        ERROR_LOCATION,
+                                        &written);
+         if (error)
+            return error;
+      }
+   }
+
+   // write message close (-1) even if we failed to retrieve pids above
+   // this prevents the child from being stuck in limbo or interpreting its
+   // actual stdin as fds
+   int close = -1;
+   Error closeError = posixCall<std::size_t>(boost::bind(::write,
+                                                         pipeFd,
+                                                         &close,
+                                                         4),
+                                             ERROR_LOCATION,
+                                             &written);
+
+   if (error)
+      return error;
+
+   return closeError;
+}
+
 namespace signal_safe {
 
 namespace {
 
-void closeFileDescriptorsFromSafe(int fdStart)
+// worst case scenario - close all file descriptors possible
+// this can be EXTREMELY slow when max fd is set to a high value
+void closeFileDescriptorsFromSafe(unsigned int fdStart, rlim_t fdLimit)
 {
    // safe function is best effort - swallow all errors
    // this is necessary when invoked in a signal handler or
    // during a fork in multithreaded processes to prevent hangs
 
-   // get limit
-   struct rlimit rl;
-   if (::getrlimit(RLIMIT_NOFILE, &rl) < 0)
-      return;
-
-   if (rl.rlim_max == RLIM_INFINITY)
-      rl.rlim_max = 1024; // default on linux
+   if (fdLimit == RLIM_INFINITY)
+      fdLimit = 1024; // default on linux
 
    // close file descriptors
-   for (int i=fdStart; i< (int)rl.rlim_max; i++)
+   for (unsigned int i = fdStart; i < fdLimit; ++i)
    {
       ::close(i);
    }
 }
 
 } // anonymous namespace
 
-void closeNonStdFileDescriptors()
+void closeNonStdFileDescriptors(unsigned int fdLimit)
+{
+   closeFileDescriptorsFromSafe(STDERR_FILENO+1, fdLimit);
+}
+
+void closeFileDescriptorsFromParent(int pipeFd, unsigned int fdStart, rlim_t fdLimit)
 {
-   return closeFileDescriptorsFromSafe(STDERR_FILENO+1);
+   // read fds that we own from parent process pipe until we've read them all
+   // the parent must give us this list because we cannot fetch it ourselves
+   // in a signal-safe way, but we can read from the pipe safely
+   bool readDescriptors = false;
+
+   char buffer[4];
+   while (true)
+   {
+      ssize_t bytesRead = ::read(pipeFd, buffer, 4);
+
+      // check for error
+      if (bytesRead == -1 || bytesRead == 0)
+         break;
+
+      // determine which fd was just read from the parent
+      int res = (int)(*buffer);
+      if (res == -1)
+         break; // indicates no more fds are open by the process
```

Comment:
Does reading 0 fds from the parent indicate that we need to close every descriptor manually? (seems like we'd fall back to closing all descriptors in this case since `readDescriptors` isn't set yet)

Reply:
That's correct. Reading 0 fds is an error condition, since at the minimum the std streams should be reported. The error handling condition is to basically fall back on the legacy behavior for closing all nonstd streams.

Reply:
Got it, thanks!